server/server.go

```diff
@@ -164,6 +165,44 @@ func (s *Server) withRecovery(next http.Handler) http.Handler {
 	})
 }
 
+func (s *Server) withValidation(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		w = newWrappedWriter(w)
+
+		defer func() {
+			elapsed := float64(time.Since(start)) / float64(time.Second)
```

Comment:
Oh I actually misunderstood you. I thought you meant that _all_ metrics code will be moved to middleware. That's why I said it can be called from a lot of places. Now I realize you were just talking about request duration. In that case, I think a separate middleware is fine. Even the name `WithValidation` is misleading because it does more than one thing. Apologies for the double work.

Let's also move the `withValidation` method inside validate_signature.go and un-export `ValidateSignature`.

And this is a great method to write some tests since this has become more modular. Let's add them too.